### ISSUE TITLE
Positional arguments can't be before flags

### ISSUE THREAD

**Issue by abourget** [alecthomas/kingpin#29]
I want to be able to have:

`executable build myproject:123 <flags>`

right now, the `myproject:123` is registered as an `Arg()` and is always placed at the end of the flags. It makes no sense for the users of my CLI tool to put the project name at the end of the flags. I need to at least support it at the beginning, if not anywhere when considered as a positional argument.

Is that possible, how complex would it be to implement ?  I was thinking at an option on the `Command()` to mark it as not having any sub-commands, so it can handle positional arguments before the flags.  What do you think ?

**Comment by alecthomas** [comment]
This is already implemented in the [v2-unstable](https://github.com/alecthomas/kingpin/tree/v2-unstable) branch. See also #24.

You can import it as `gopkg.in/alecthomas/kingpin.v2-unstable`.

**Comment by alecthomas** [comment]
PS. If you find a case where this doesn't work with v2-unstable, it's a bug.

**Comment by abourget** [comment]
ohh sweet thanks :)

**Comment by abourget** [comment]
I have:

```
	download           = app.Command("download", "Retrieve a build Artifact")
	downloadProjectRev = download.Arg("projectRev", "Project build ID to get. 'project:1001' or 'project' for latest.").Required().String()
	downloadOutput     = download.Flag("output", "Filename where to write content").Short('o').PlaceHolder("FILE").Required().String()
```
with `v2-unstable` doesn't allow me to run:

```
$ alchemist download test:1001 -o superbob.txt
alchemist: error: required flag --output not provided, try --help
```

I understand ambiguity between the next command and a positional argument after a command, so maybe we could disable "sub-commands" on a certain command, to alleviate that confusion.

So unless I missed something on the `v2-unstable` branch, it doesn't work as I expect.

**Comment by abourget** [comment]
so I'm not talking about ordering of flags (like #24 does), but ordering of positional arguments.

**Comment by abourget** [comment]
hmm, actually, you probably check for subcommands before doing any positional arguments filling, so there wouldn't even be any confusion.

What I mean is similar to Python's treatment of positional versus keyword arguments..  positional arguments are required to appear before the keyword arguments, so positional argument 1 is first, arg 2 is second, etc.. If you put a keyword argument in between, then they cannot be aligned to their definition anymore. (positional argument 3 can appear 4th!).

**Comment by alecthomas** [comment]
There is currently a bug preventing it working. #30 from memory.

**Comment by alecthomas** [comment]
v2-unstable (once the bug is fixed) will allow:

```
alchemist download test:10001 -o superbob.txt
```

However, this will also work:

```
alchemist download -o superbob.txt test:10001
```

This is commonly called [interspersed flags](https://docs.python.org/3.0/library/optparse.html#querying-and-manipulating-your-option-parser), which optparse also supports.

**Comment by alecthomas** [comment]
Okay, this should be fixed in 9315ed681d2d4dd60507d156897afc6f0b7ec347.
